### PR TITLE
[otbn] Use .+123 syntax where possible for relative operands

### PR DESCRIPTION
**Pull request by rswarbrick**
The binutils RISC-V assembler does magic things for instructions like

    bne x0, x1, 256

and expands them to something like:

    beq x0, x1, .+8
    jal 256

This plays havoc with the random instruction generator, but can be
avoided by explicitly writing

    bne x0, x1, .+128

(assuming the PC was 128).

### CODE REVIEW

File: hw/ip/otbn/util/otbn-as [hunk 1]
```diff
@@ -990,6 +990,7 @@ class Transformer:
 
         match = insn.asm_pattern.match(operands_str.rstrip())
         if match is None:
+            print(insn.asm_pattern)
```
Comment (imphil) [inline hw/ip/otbn/util/otbn-as:993]: Intentional or debugging aid?

File: hw/ip/otbn/util/shared/operand.py [hunk 2]
```diff
@@ -305,6 +303,12 @@ def str_to_op_val(self, as_str: str) -> Optional[int]:
         except ValueError:
             return None
 
+    def op_val_to_str(self, op_val: int, cur_pc: Optional[int]) -> str:
+        if self.pc_rel and cur_pc is not None:
+            return '.+{}'.format(op_val - cur_pc)
```
Comment (imphil) [inline hw/ip/otbn/util/shared/operand.py:308]: Maybe add a comment on (the need for) this "magic" way of writing immediates?

File: hw/ip/otbn/util/shared/syntax.py [hunk 3]
```diff
@@ -68,15 +68,24 @@ def asm_pattern(self) -> str:
         # We also split out ++ and -- separately, to disambiguate things like
         # x1++, which must be parsed as x1 followed by ++.
         #
+        # However, we do want to allow things like ".+123". To get this right,
+        # suppose C matches any character other than elements of " ,+-". Then
```
Comment (imphil) [inline hw/ip/otbn/util/shared/syntax.py:72]:
```suggestion
        # suppose S matches any character other than elements of " ,+-". Then
```

?

**Review by imphil — APPROVED** [review]
LGTM with comments.

@asb The assembly syntax matches what you described this morning, doesn't it?

**Comment by asb** [comment]
> LGTM with comments.
> 
> @asb The assembly syntax matches what you described this morning, doesn't it?

Yep, this is what I was talking about. Way back in 2018 I kicked off a discussion in the LLVM community about how this behaviour differs for different targets between LLVM and GCC. Peter Smith (Arm/Linaro) [commented](https://lists.llvm.org/pipermail/llvm-dev/2018-January/120331.html) that from his perspective a branch to a bare number is effectively implementation defined for the Arm and AArch64 assemblers. I suppose other architectures take a similar view.

File: hw/ip/otbn/util/otbn-as (hunk 1 above)
Comment (rswarbrick) [reply hw/ip/otbn/util/otbn-as:993]: Oops! Thanks.

File: hw/ip/otbn/util/shared/syntax.py (hunk 3 above)
Comment (rswarbrick) [reply hw/ip/otbn/util/shared/syntax.py:72]: Ah, yes, thanks. I couldn't decide what to call it and ended up with S for **s**ymbol constituent.

**Pull request merged**